Fw 956 refactor on changes

benlesh opened the pull request:
This change inverts the control over who is setting up onChanges to fire from the parent template to the component/directive definition itself. The idea is that the new `NgOnChangesFeature` will add a `setInput` function to the directive definition that is to be called when setting an input from a parent template. That `setInput` function will be composed to update a `SimpleChanges` object that is now stored at a known private property on the component instance.

Reverts most of the changes from the #27965 PR that moved control of onChanges into the parent template's instructions.

benlesh committed:
refactor(ivy): revert onChanges change back to a feature

- adds fixmeIvy annotation to tests that should remain updated so we can resolve those issues in the subsequent commits

benlesh committed:
refactor(ivy): Add newer, smaller NgOnChangesFeature

mary-poppins commented:
You can preview 9cff51c at https://pr28187-9cff51c.ngbuilds.io/.

mary-poppins commented:
You can preview c39d3c7 at https://pr28187-c39d3c7.ngbuilds.io/.

kara commented on packages/upgrade/test/static/integration/downgrade_module_spec.ts:
```diff
@@ -722,63 +722,66 @@ withEachNg1Version(() => {
 
-      it('should propagate input changes inside the Angular zone', async(() => {
-           let ng2Component: Ng2Component;
+      fixmeIvy('FW-715: ngOnChanges being called a second time unexpectedly')
```
Seems like this was left in here by mistake? This should not fail anymore. Same below

kara commented on packages/upgrade/test/static/integration/downgrade_component_spec.ts:
```diff
@@ -22,104 +22,106 @@ withEachNg1Version(() => {
-           $rootScope['eventA'] = '?';
-           $rootScope['eventB'] = '?';
-         });
+    fixmeIvy('FW-715: ngOnChanges being called a second time unexpectedly')
```
These fixmes should also be removed.

kara commented on packages/upgrade/test/dynamic/upgrade_spec.ts:
```diff
@@ -316,235 +316,239 @@ withEachNg1Version(() => {
-             }
-           }
-           ng1Module.directive('ng2', adapter.downgradeNg2Component(Ng2));
+      fixmeIvy('FW-715: ngOnChanges being called a second time unexpectedly')
```
Same here, these fixmes should be removed.

kara commented on packages/core/test/render3/ng_on_changes_feature_spec.ts:
```diff
@@ -0,0 +1,18 @@
+import {DirectiveDef, NgOnChangesFeature, defineDirective} from '../../src/render3/index';
+
+describe(
+    'NgOnChangesFeature', () => {
```
Seems like we should remove this file if there aren't any tests in it.

benlesh replied:
Right... yeah. I left it because I thought I might add some, but the lifecycle_spec seems to cover everything.

kara commented on packages/core/test/bundling/hello_world/bundle.golden_symbols.json:
```diff
@@ -81,13 +81,13 @@
   },
   {
-    "name": "NodeInjectorFactory"
+    "name": "NgOnChangesFeature"
```
It seems odd that we are bringing `NgOnChangesFeature` into `hello_world`. Why would that be the case?

benlesh replied:
Suspect this is a bug with the how the compiler is identifying the presence of the feature and adding it. I think we can fix this in a separate PR.

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1186,18 +1186,29 @@ export function createTNode(
+    ngDevMode && assertDataInRange(lView, index);
+    const def = tView.data[index] as DirectiveDef<any>;
+    const setInput = def.setInput;
+    if (setInput) {
```
The intermediate assignment seems unnecessary, removing it avoids the `!`:
```typescript
if (def.setInput) {
  def.setInput(instance, value, publicName, privateName);
}
```

benlesh replied:
It's an optimization to prevent it from being looked up twice.

JoostK replied:
@benlesh it would be, if the call didn't rely on `this` being the `DirectiveDef` (or we should use `setInput.call(def, …)`. We are currently reading it twice already because of that.

kara commented on packages/compiler/src/render3/view/api.ts:
```diff
@@ -74,6 +74,17 @@ export interface R3DirectiveMetadata {
+  lifecycle: {
+    /**
+     * Whether the directive uses NgOnChanges.
+     */
```
Nit: comment indentation

benlesh replied:
:) Clang format did that... it's because it's inside of an indented block (part of the `lifecycle: { }` object.

kara commented on packages/core/src/interface/lifecycle_hooks.ts:
```diff
@@ -5,9 +5,20 @@
  * found in the LICENSE file at https://angular.io/license
  */
-import {SimpleChanges} from './simple_change';
+import {SimpleChange, SimpleChanges} from './simple_change';
```
Seems like we don't need to import `SimpleChanges` since we are redefining it?

benlesh replied:
We shouldn't be redefining it here. I'll fix that.

kara reviewed (changes requested):
General approach looks good, but we need the tests turned on to verify

JoostK commented on packages/core/src/render3/features/ng_onchanges_feature.ts:
```diff
@@ -40,86 +40,64 @@ type OnChangesExpando = OnChanges & {
+function ngOnChangesSetInput<T>(
+    this: DirectiveDef<T>, instance: T, value: any, publicName: string, privateName: string): void {
+  const simpleChangesStore = getSimpleChangesStore(instance) ||
+      setSimpleChangesStore(instance, {previous: null, current: null});
```
Seems as if we can assign `EMPTY_OBJ` into `previous` here immediately, such that it is guaranteed to be present (the `NgSimpleChangesStore.previous` property may loose `null` from its type and the assignment of local var `previous` below can be inlined into the assignment of `previousChange`)

JoostK reviewed (commented):
Yeah! 🎉

benlesh committed:
fixup! refactor(ivy): Add newer, smaller NgOnChangesFeature

mary-poppins commented:
You can preview 8c07bd4 at https://pr28187-8c07bd4.ngbuilds.io/.

benlesh commented:
Presubmit: http://test/OCL:229648942:BASE:229649137:1547685156659:dab1c888

benlesh committed:
fixup! refactor(ivy): Add newer, smaller NgOnChangesFeature

mary-poppins commented:
You can preview 00bc640 at https://pr28187-00bc640.ngbuilds.io/.

kara reviewed (approved):
LGTM.

benlesh commented:
presubmit: http://test/OCL:229648942:BASE:229853217:1547777095425:d64118f5

alxhub commented:
Note: missing reviews.

benlesh commented on packages/common/src/directives/ng_template_outlet.ts:
```diff
@@ -34,20 +34,14 @@ import {Directive, EmbeddedViewRef, Input, OnChanges, SimpleChange, SimpleChange
 @Directive({selector: '[ngTemplateOutlet]'})
 export class NgTemplateOutlet implements OnChanges {
-  private _viewRef: EmbeddedViewRef<any>|null = null;
+  // TODO(issue/24571): remove '!'.
```
This file appears to have been accidentally reverted when I reverted the other onChanges changes.

(changes changes changes)

benlesh committed:
fixup! refactor(ivy): Add newer, smaller NgOnChangesFeature

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -174,6 +174,11 @@ export function extractDirectiveMetadata(
+  // Determine if `ngOnChanges` is a lifecycle hook defined on the component.
+  const usesOnChanges = members.some(
+      member => !member.isStatic && member.kind === ClassMemberKind.Method &&
+          member.name === 'ngOnChanges');
```
Does this account for inherited `ngOnChanges`?

petebacondarwin reviewed (approved):
There do not appear to be any ngUpgrade files left in this PR?

alxhub reviewed (approved):
Compiler-wise LGTM

benlesh commented:
Oops, broken symbol test... I'll get that fixed here in a jiff.

benlesh committed:
fixup! refactor(ivy): Add newer, smaller NgOnChangesFeature

benlesh committed:
fixup! refactor(ivy): Add newer, smaller NgOnChangesFeature

benlesh committed:
fix(ivy): NgOnChangesFeature no longer included in hello_world

- Wraps the NgOnChangesFeature in a factory such that no side effects occur in the module root
- Adds comments to ngInherit property on feature definition interface to help guide others not to make the same mistake
- Updates compiler to generate the feature properly after the change to it being a factory
- Updates appropriate tests

benlesh commented:
Adding one more bit of work to this to resolve FW-965, which is that the `NgOnChangesFeature`, even as it previously existed, was being included in the hello_world bundle, which makes no reference to it.

benlesh committed:
fixup! fix(ivy): NgOnChangesFeature no longer included in hello_world

benlesh commented:
presubmit: http://test/OCL:229648942:BASE:230378383:1548185415508:de230301

kara commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -302,11 +302,27 @@ export type PipeDefWithMeta<T, Name extends string> = PipeDef<T>;
+   *
+   * NOTE: DO NOT SET IN ROOT OF MODULE! Doing so will result in tree-shakers/bundlers
+   * identifying the change as a side effect, and the feature will be included in
+   * every bundle.
```
Thanks for including this 👍

kara reviewed (approved):
LGTM

kara commented:
merge assistance: no longer need public-API approval because the public API changes were reverted (no longer in this PR). Should be good with approval from @mhevery and i.
